main/src/library/File.flix

```diff
@@ -59,4 +165,196 @@ namespace File {
+                import java.lang.Throwable.getMessage();
+                Err(getMessage(ex))
+        }
+
+
+    ///
+    /// Helper function for read.
+    /// Converts a list of strings to a string with linebreaks between every string in `list`.
+    ///
+    def readHelper(list: List[String]): String & Pure =
+        List.foldLeft((x, y) -> x + "\n" + y, "", list)
+
+
+    ///
+    /// Returns a string containing the given file `f` with the options `opts`.
+    ///
+    pub def readWith(opts: {offset :: Int64, count :: Int32, charSet :: String}, file : String): Result[String, String] & Impure = {
+        try {
+            import new java.io.File(String) as newFile;
+            import java.nio.charset.Charset:forName(String);
+            import java.io.File.toPath();
+            import java.nio.file.Files:newBufferedReader(##java.nio.file.Path, ##java.nio.charset.Charset);
+            let javaFile = newFile(file);
+            let javaPath = toPath(javaFile);
+            let reader = newBufferedReader(javaPath, forName(opts.charSet));
+            let fileRead = readHelper(readAll(reader));
+            Ok(readWithHelper(fileRead, opts.offset, opts.count))
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage();
+                Err(getMessage(ex))
+        }
+    }
+
+
+    ///
+    /// Helper function for readWith.
+    /// Takes a String and applys the options.
+    ///
+    def readWithHelper(file: String, offset: Int64, count: Int32): String & Pure =
+        let fileLength = String.length(file);
+
+        match Int64.tryToInt32(offset) {
+            case None => String.dropRight(fileLength - count, file)
+            case Some(n) =>
+                let str = String.dropLeft(n, file);
+                let strLength = String.length(str);
+                String.dropRight(strLength - count, str)
+        }
+
+
+    ///
+    /// Returns an iterator of the given `file`
+    ///
+    pub def readLinesIter(file: String): Iterator[Result[String, String]] & Impure =
+        try {
+            import new java.io.File(String) as newFile;
+            import java.io.File.toPath();
+            import java.nio.file.Files:newBufferedReader(##java.nio.file.Path);
+            import java.io.BufferedReader.readLine();
+
+            let javaFile = newFile(file);
+            let javaPath = toPath(javaFile);
+            let reader = newBufferedReader(javaPath);
+            let line = ref readLine(reader);
+
+            let done = () -> Object.isNull(deref line);
+            let next = () -> {
+                let l = deref line;
+                line := readLine(reader);
+                Ok(l)
+            };
+            Iterator(done, next)
+        } catch {
+            case ex: ##java.io.IOException =>
+                import java.lang.Throwable.getMessage();
+                let err = ref getMessage(ex);
+                let done = () -> Object.isNull(deref err);
```

Comment:
`deref err` is never going to be `null`, right? I think `err` can be a regular String and then we could use a `Ref[Bool]` to indicate whether the iterator is done (just to avoid null whenever possible)

Reply:
Actually, is there an `Iterator.singleton` function? (If not, then there should be)

Reply:
Something like this instead maybe?
`case ex: ##java.io.IOException =>`
     `import java.lang.Throwable.getMessage();`
     `Iterator.singleton(Err(getMessage(ex)))`

Reply:
Yes, that would be ideal!